JSON: Extract register information

AbhinavMir opened the pull request:
for #10 
Adding a new PR because the previous one was getting stale to the point of being too messy to rebase

ThinkOpenly commented on src/sail_json_backend/json.ml:
```diff
@@ -98,6 +99,41 @@ let dequote qs =
 let string_of_arg = function E_aux (E_id id, _) -> "\"" ^ string_of_id id ^ "\"" | exp -> "exp " ^ string_of_exp exp
 
+(* Function to parse the number at the end of a register name *)
+let parse_reg_name name =
+  try
+    let i = String.rindex name '_' + 1 in
+    (* Find the last underscore, start reading the number just after it *)
+    int_of_string (String.sub name i (String.length name - i))
+  with _ -> -1 (* Return -1 if no number is found or the substring operation fails *)
```
For the current Sail code, this always returns -1.

AbhinavMir replied:
Based on previous conversation, I've since removed this. I was for whatever reason, expecting an underscore, and even upon not expecting that, the sorting wasn't correct. We had a similar discussion sometime back based on this and decided to not do lexographic sorting, but I still tried it out. Would we want to sort it x0 through x11 for example in a good manner?

ThinkOpenly replied:
Ideally, we'd want the human-expected sorting, but that seems harder than it's worth. I'm content with type-then-lexicographical sorting, indefinitely.

ThinkOpenly commented on src/sail_json_backend/json.ml:
```diff
@@ -98,6 +99,41 @@ let dequote qs =
+            (* If types are the same, sort by parsed number *)
+            let num1 = parse_reg_name n1 in
+            let num2 = parse_reg_name n2 in
+            if num1 = num2 then String.compare n1 n2 else compare num1 num2
```
What is `compare` here? Is it supposed to be `String.compare`?

AbhinavMir replied:
Yep! My bad!

ThinkOpenly commented on src/sail_json_backend/json.ml:
```diff
@@ -98,6 +99,41 @@ let dequote qs =
+      )
+      sorted_regs
+  in
+  let reg_list = String.concat "" json_regs in
+  let final_reg_list = if reg_list = "" then "" else String.sub reg_list 0 (String.length reg_list - 1) in
+  "[ " ^ final_reg_list ^ "\n]"
```
I don't think these are used anywhere.

AbhinavMir replied:
Remains of an unused code I forgot to remove

ThinkOpenly reviewed (changes requested):
If I've jumped the gun, and this is not ready for review, I apologize. Feedback inline, but a general comment is that you seem to have code that tries to sort a list of "numbered identifiers" like "x0, x1, x2, x3, ..., x10, x11, ..." as a human might expect, rather than simply lexicographically ("x0, x1, x10, x2, ..."). However, you are expecting an underscore to separate the alpha part of the identifier from the numeric part, and that apparently never happens. And some identifiers embed the number within, like "pmp12cfg". We may want to go without trying to ferret those numbers out for sorting for now, and just go with a sort based first on register type, which you have already, and then lexicographially within those types.

AbhinavMir commented:
@ThinkOpenly Indeed I marked this as ready for review when it should've been in draft, I've taken your inputs and made some modifications. Please let me know if they're satisfactory. I've removed sorting and unused variables. I had switched to using a plain `vi` since VS Code was forcing it's own linter which was getting annoying and I must've missed that unused variable.

ThinkOpenly commented:
Looks good. Did you run `dune fmt`?

ThinkOpenly commented:
The "Close with comment" button is very close to the "Comment" button. Sorry about that. Check formatting using `dune fmt`, then let's get this merged.

ThinkOpenly commented:
Oh! I guess the GitHub actions run `dune fmt` automatically. Duh. So, ignore my previous comments. I'll wait for the checks to complete, then merge this.

ThinkOpenly commented on src/sail_json_backend/json.ml:
```diff
@@ -98,6 +99,21 @@ let dequote qs =
+  in
+  let json_regs =
+    List.map
+      (fun (name, regtype) ->
+        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
+      )
+      sorted_regs
+  in
+  "[ " ^ String.concat "" json_regs ^ "\n]"
```
Actually, this has a similar problem to one I just fixed.

The emitted JSON for this content ends like:
```
  {
    "name": "vlenb",
    "type": "xlenbits"
  },
]
```
...it has a comma after the last element, which the Python JSON module does not like to load.

I think there's an easy fix: remove the comma at the end of line 111, and add it inside the empty string on line 115.

Here's how to test for valid JSON:
```
python3 -c "import json; f=open('AbhinavMir.out'); j=json.load(f)"
```

However, you'll need my recent fix (or delete the dangling comma at the end of the "functions" content). A rebase to my JSON branch should pull in the fix.

AbhinavMir replied:
Shold I just remove the comma or add it in line 115 to make it merge easy with latest code?

AbhinavMir replied:
Thank you Paul. Can I request a suggestion here? When I try to do the current suggested fix or removing a `,` and adding it back in on a later line, it gives me a still invalid JSON.

```ocaml
let json_of_registers () =
  let regs = Hashtbl.fold (fun name regtype accum -> (name, regtype) :: accum) registers [] in
  let sorted_regs =
    List.sort (fun (n1, t1) (n2, t2) -> match String.compare t1 t2 with 0 -> String.compare n1 n2 | c -> c) regs
  in
  let json_regs =
    List.map
      (fun (name, regtype) ->
        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
      )
      sorted_regs
  in
  "[ " ^ String.concat "" json_regs ^ "\n]"
```

ThinkOpenly replied:
This works for me:
```
   let json_regs =
     List.map
       (fun (name, regtype) ->
-        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  },"
+        "\n  {\n" ^ "    \"name\": \"" ^ name ^ "\",\n" ^ "    \"type\": \"" ^ regtype ^ "\"\n" ^ "  }"
       )
       sorted_regs
   in
-  "[ " ^ String.concat "" json_regs ^ "\n]"
+  "[ " ^ String.concat "," json_regs ^ "\n]"
```

With your current branch state, that doesn't include 7ad136c5b2ba1e9c1cf9fef911a20173b7394042, you need to edit the emitted JSON to remove the dangling comma after the "functions" content:
```
  {
    "name": "riscv_f32Div",
    "source": "{\n  extern_f32Div(rm, v1, v2);\n  (float_fflags[4 .. 0], float_result[31 .. 0])\n}"
  }, (* <-- remove this comma *)
  ]
}
(* EOF *)
```

AbhinavMir replied:
I just dealt with that but it seems like there's still an unexpected `,` being generated towards the end of functions in the JSON. Is this the problem you dealt with right now but isn't merged into the code I forked? Just making sure.
